File: realm-common/src/main/java/io/github/solusmods/eternalcore/realm/impl/network/c2s/RequestRealmBreakthroughPacket.java
package io.github.solusmods.eternalcore.realm.impl.network.c2s;

import dev.architectury.networking.NetworkManager;
import dev.architectury.utils.Env;
import io.github.solusmods.eternalcore.realm.EternalCoreRealm;
import io.github.solusmods.eternalcore.realm.api.Realm;
import io.github.solusmods.eternalcore.realm.api.RealmAPI;
import io.github.solusmods.eternalcore.realm.api.RealmInstance;
import io.github.solusmods.eternalcore.realm.api.Realms;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.player.Player;
import org.jetbrains.annotations.NotNull;

import java.util.Optional;

public record RequestRealmBreakthroughPacket(
        ResourceLocation realm
) implements CustomPacketPayload {
    public static final Type<RequestRealmBreakthroughPacket> TYPE = new Type<>(EternalCoreRealm.create("request_realm_breakthrough"));
    public static final StreamCodec<FriendlyByteBuf, RequestRealmBreakthroughPacket> STREAM_CODEC = CustomPacketPayload.codec(RequestRealmBreakthroughPacket::encode, RequestRealmBreakthroughPacket::new);

    public RequestRealmBreakthroughPacket(FriendlyByteBuf buf) {
        this(buf.readResourceLocation());
    }

    public void encode(FriendlyByteBuf buf) {
        buf.writeResourceLocation(this.realm);
    }

    public void handle(NetworkManager.PacketContext context) {
        if (context.getEnvironment() != Env.SERVER) return;
        context.queue(() -> {
            Player player = context.getPlayer();
            if (player == null) return;

            Realms storage = RealmAPI.getRealmFrom(player);
            Optional<RealmInstance> optional = storage.getRealm();
            if (optional.isEmpty()) return;

            Realm realm1 = RealmAPI.getRealmRegistry().get(realm);
            if (realm1 == null) return;

            RealmInstance instance = optional.get();
            if (!instance.getNextBreakthroughs(player).contains(realm1)) {
            }

//            storage.evolveRace(realm1);
        });
    }

    public @NotNull Type<RequestRealmBreakthroughPacket> type() {
        return TYPE;
    }
}

File: realm-common/src/main/kotlin/io/github/solusmods/eternalcore/realm/impl/network/c2s/RequestRealmBreakthroughPacket.kt
package io.github.solusmods.eternalcore.realm.impl.network.c2s

import dev.architectury.networking.NetworkManager
import dev.architectury.utils.Env
import io.github.solusmods.eternalcore.realm.EternalCoreRealm
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.codec.StreamDecoder
import net.minecraft.network.codec.StreamMemberEncoder
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.ResourceLocation


data class RequestRealmBreakthroughPacket(
    val realm: ResourceLocation?
) : CustomPacketPayload {
    constructor(buf: FriendlyByteBuf) : this(buf.readResourceLocation())

    fun encode(buf: FriendlyByteBuf) {
        buf.writeResourceLocation(this.realm)
    }

    fun handle(context: NetworkManager.PacketContext) {
        if (context.getEnvironment() != Env.SERVER) return
        context.queue(Runnable { ClientAccess.handle(this, context.getPlayer()) })
    }

    override fun type(): CustomPacketPayload.Type<RequestRealmBreakthroughPacket?> {
        return TYPE
    }

    companion object {
        val TYPE: CustomPacketPayload.Type<RequestRealmBreakthroughPacket?> =
            CustomPacketPayload.Type<RequestRealmBreakthroughPacket?>(EternalCoreRealm.create("request_realm_breakthrough"))
        val STREAM_CODEC: StreamCodec<FriendlyByteBuf?, RequestRealmBreakthroughPacket?> =
            CustomPacketPayload.codec<FriendlyByteBuf?, RequestRealmBreakthroughPacket?>(
                { obj: RequestRealmBreakthroughPacket?, buf: FriendlyByteBuf? -> obj!!.encode(buf!!) },
                { buf: FriendlyByteBuf? ->
                    RequestRealmBreakthroughPacket(
                        buf!!
                    )
                })
    }
}
